core/cas-server-core-authentication-api/src/main/java/org/apereo/cas/authentication/AuthenticationCredentialsThreadLocalBinder.java

```diff
@@ -24,12 +27,31 @@
 
     private static final ThreadLocal<Authentication> CURRENT_AUTHENTICATION = new ThreadLocal<>();
 
+    private static final ThreadLocal<Principal> IN_PROGRESS_PRINCIPAL = new ThreadLocal<>();
+
     private static final ThreadLocal<Authentication> IN_PROGRESS_AUTHENTICATION = new ThreadLocal<>();
 
     private static final ThreadLocal<AuthenticationBuilder> CURRENT_AUTHENTICATION_BUILDER = new ThreadLocal<>();
 
     private static final ThreadLocal<String[]> CURRENT_CREDENTIAL_IDS = new ThreadLocal<>();
 
+    /**
+     * Bind Principal to ThreadLocal for principal that has been authenticated but for which attributes have not yet been retrieved.
+     * This can make attributes from the authentication event (e.g. X509, LDAP, WSFED) available to scripted attribute repositories.
+     * If a principal is already bound, combine the attributes with new attributes overriding existing.
```

Comment:
Nitpicking here: *to scripted attribute repositories* is a little too specific. I might suggest that the description here be generalized a bit more; something like "to consumer components that may not have immediate access to the resolved principal, such as ...*